Add a blog post about QueueingHint (#43686)
kubernetes/website: content/en/blog/_posts/2023-11-xx-scheduler-queueinghint/index.md

---
layout: blog

> title: "Kubernetes v1.28: QueueingHint brings a new possibility to optimize our scheduling"

Reviewer: 1.29 now, please. Also - until we change the style guide: [Title Case](https://en.wikipedia.org/wiki/Title_case).

Author: Good to know.

Reviewer: I've marked this conversation as “not resolved” because the feedback is still applicable.

> date: 2023-10-25T10:00:00-08:00

Author: TODO: update the date here + the folder name once the public date is decided.

slug: scheduler-queueinghint
---

**Author:** [Kensei Nakada](https://github.com/sanposhiho) (Mercari)

The scheduler is the core component that decides which Node Pods run on.
Basically, it schedules Pods **one by one**, 
and thus the larger your cluster is, the more crucial the throughput of the scheduler is.

The throughput of the scheduler is our eternal challenge, 

> over the years, SIG-Scheduling have been putting effort to improve the scheduling throughput by many enhancements.

Reviewer: ```suggestion
over the years, SIG Scheduling have been putting effort to improve the scheduling throughput by many enhancements.
```

> In this blog post, I'll introduce a recent major improvement in the scheduler, named QueueingHint.
> 
> We'll go through the explanation of the basic background knowledge of the scheduler,
> and how QueueingHint improves our scheduling throughput.

Reviewer: How about:
```suggestion
In this blog post, I'll introduce a recent major improvement in the scheduler: a new
[scheduling context element](/docs/concepts/scheduling-eviction/scheduling-framework/#extension-points)
named _QueueingHint_.

We'll go through the explanation of the basic background knowledge of the scheduler,
and how QueueingHint improves our scheduling throughput.
```

> ## Scheduling Queue

Reviewer: ```suggestion
## Scheduling queue
```

> The scheduler has Scheduling Queue which has all unscheduled Pods.

Reviewer 2: ```suggestion
The scheduler stores all unscheduled Pods in an internal component that we call the Scheduling Queue.
```

> Scheduling Queue is composed of three places in it - ActiveQ, BackoffQ and Unschedulable Pod Pool.

Reviewer 2: ```suggestion
The Scheduling Queue is composed of three data structures: ActiveQ, BackoffQ and Unschedulable Pod Pool.
```

> - ActiveQ: Pods which are ready to get scheduling.

Reviewer 2: ```suggestion
- ActiveQ: It holds newly created Pods or Pods which are ready to be retried for scheduling.
```

> - BackoffQ: Pods which are waiting for the backoff, and will be put into ActiveQ after that.

Reviewer 2: ```suggestion
- BackoffQ: It holds Pods which are ready to be retried, but are waiting for a backoff period, which depends on the number of times the scheduled attempted to schedule the Pod.
```

> - Unschedulable Pod Pool: Pods which should not be scheduled for now.

Reviewer 2: ```suggestion
- Unschedulable Pod Pool: It holds Pods which should not be scheduled for now, because they have a Scheduling Gate or because the scheduler attempted to schedule them and nothing has changed in the cluster that could make the Pod schedulable.
```

> ## Scheduling Framework and Plugins
> 
> [Scheduling Framework](/docs/concepts/scheduling-eviction/scheduling-framework/)
> 
> The scheduler is implemented with Scheduling Framework. 

Reviewer: ```suggestion
## Scheduling framework and plugins

The Kubernetes scheduler is implemented following the Kubernetes
[scheduling framework](/docs/concepts/scheduling-eviction/scheduling-framework/).
```

> And, each scheduling requirements/preferences is implemented as a plugin.

Reviewer 2: ```suggestion
And, each scheduling requirements are implemented as a plugin.
```

> (e.g., PodAffinity is implemented in the PodAffinity plugin.)

Reviewer: ```suggestion
(e.g., [Pod affinity](/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity)
is implemented in the `PodAffinity` plugin.)
```

> The first phase, called Scheduling Cycle, takes Pods from activeQ **one by one**, gather all plugins' idea,

Reviewer 2: ```suggestion
The first phase, called the Scheduling Cycle, takes Pods from activeQ **one by one**, runs all plugins' logic,
```

> and lastly decides a Node to run the Pod, or concludes that the Pod cannot go to anywhere for now.

Reviewer 2: ```suggestion
and lastly decides in which Node to run the Pod, or concludes that the Pod cannot go to anywhere for now.
```

> If the scheduling is successful, the second phase, called Binding Cycle, binds the Pod with the Node.

Reviewer 2: ```suggestion
If the scheduling is successful, the second phase, called Binding Cycle, binds the Pod with the Node by communicating the decision to the apiserver.
```

But, if it turns out that the Pod cannot go to anywhere in Scheduling Cycle, 
Binding Cycle isn't executed, instead the Pod is moved back to Scheduling Queue.

> There are some exception cases though, such unscheduled Pod is basically put into Unschedulable Pod Pool.

Reviewer 2: ```suggestion
Although there are some exceptions, unscheduled Pods enter the Unschedulable Pod Pool.
```

Pods in Unschedulable Pod Pool are moved to ActiveQ/BackoffQ 

> only when Scheduling Queue thinks they might be schedulable if we retry the scheduling.

Reviewer 2: instead of think, we can say "identify changes in the cluster"

That is a crucial step because Scheduling Cycle is performed for Pods one by one -
if we didn't have Unschedulable Pod Pool and kept retrying the scheduling of any Pods, 

> Scheduling Cycle is wasted for Pods with no hope to be scheduled.

Reviewer 2: ```suggestion
multiple Scheduling Cycles would be  wasted for Pods with no chance to be scheduled.
```

> Then, how do they decide when to move? How do they notice that Pods might be schedulable now?

Reviewer 2: ```suggestion
Then, how do they decide when to move a Pod back into the ActiveQ? How do they notice that Pods might be schedulable now?
```

> There we go, QueueingHint comes in.

Reviewer 2: ```suggestion
Here QueueingHints come into play.
```

## QueueingHint

> QueueingHint is callback functions per plugin to notice an object addition/update/deletion in the cluster (we call them cluster events)

Reviewer 2: ```suggestion
QueueingHint is callback function per plugin to notice an object addition/update/deletion in the cluster (we call them cluster events)
```

that may make Pods schedulable.

> Let's say PodA has a required PodAffinity, and got rejected in scheduling cycle by PodAffinity plugin 
> because no Node has any Pod matching with PodA's PodAffinity.

Reviewer: ```suggestion
Let's say the Pod `pod-a` has a required Pod affinity, and got rejected in scheduling cycle by the `PodAffinity` plugin
because no Node has any Pod matching the Pod affinity specification for `pod-a`.
```

![PodA got rejected by PodAffinity](./queueinghint1.png)

> When an unscheduled Pod is put into Unschedulable Pod Pool, Scheduling Queue remembers which plugins caused the scheduling failure of the Pod.

Reviewer 2: ```suggestion
When an unscheduled Pod is put into Unschedulable Pod Pool, Scheduling Queue records which plugins caused the scheduling failure of the Pod.
```

In this example, Scheduling Queue notes that PodA was rejected by PodAffinity.

> PodA will never be schedulable until PodAffinity failure is resolved somehow.

Reviewer 2: ```suggestion
PodA will never be schedulable until the PodAffinity failure is resolved somehow.
```

> Scheduling Queue uses QueueingHint from failure plugins, which is PodAffinity in the example.

Reviewer 2: ```suggestion
The Scheduling Queue uses the QueueingHints from plugins that rejected the Pod, which is PodAffinity in the example.
```

> QueueingHint subscribes a perticular cluster event and make a decision whether an incoming event could make the Pod schedulable.

Reviewer 2: ```suggestion
A QueueingHint subscribes to a particular kind of cluster event and make a decision whether an incoming event could make the Pod schedulable.
```

Thinking about when PodAffinity failure could be resolved,
one possible scenario is that an existing Pod gets a new label which matches with PodA's PodAffinity.

PodAffinity QueueingHint checks all Pod updates happening in the cluster, 
and when it catches such update, the scheduling queue moves PodA to activeQ/backoffQ.

![PodA is moved by PodAffinity QueueingHint](./queueinghint2.png)

> ## What's new in v1.28

Reviewer 2: Maybe mention that we had an earlier mechanism before, the prechecks. And what their short-comings where.

We have been working on the development of QueueingHint since v1.27.
In v1.27, only one alpha plugin (DRA) supported QueueingHint,

> and in v1.28, some stable plugins start to work with QueueingHint.

Reviewer 2: ```suggestion
and in v1.28, some stable plugins start to implement QueueingHints.
```

> QueueingHint is not something user-facing, but we have a feature gate () as a safety net 

Reviewer 2: don't forget to add the name of the feature gate.

> because QueueingHint changes a critical path of the scheduler a lot.

Reviewer 2: ```suggestion
because QueueingHint changes a critical path of the scheduler and adds some memory overhead, depending on how busy a cluster is.
```

Author: 😅
